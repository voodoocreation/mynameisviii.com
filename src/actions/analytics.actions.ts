import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("ANALYTICS");

export const trackEvent = createAction<{
  event: string;
  [index: string]: any;
}>("TRACK_EVENT");
