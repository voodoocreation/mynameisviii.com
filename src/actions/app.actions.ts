import actionCreatorFactory from "typescript-fsa";

const createAction = actionCreatorFactory("APP");

export const initApp = createAction.async<{ locale?: string }, {}>("INIT");

export const toggleNavigation = createAction("TOGGLE_NAVIGATION");

export const setOnlineStatus = createAction<boolean>("SET_ONLINE_STATUS");

export const setHasNewVersion = createAction<boolean>("SET_HAS_NEW_VERSION");

export const setCurrentRoute = createAction<string>("SET_CURRENT_ROUTE");

export const changeRoute = createAction.async<string, {}>("CHANGE_ROUTE");
