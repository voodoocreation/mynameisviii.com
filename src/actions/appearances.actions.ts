import actionCreatorFactory from "typescript-fsa";

import { IAppearance, IDynamoResponse, ILatLng } from "../models/root.models";

const createAction = actionCreatorFactory("APPEARANCES");

export const fetchAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH");

export const fetchMoreAppearances = createAction.async<
  {},
  IDynamoResponse<IAppearance, "startingAt">
>("FETCH_MORE");

export const fetchAppearanceBySlug = createAction.async<string, IAppearance>(
  "FETCH_BY_SLUG"
);

export const setCurrentAppearanceSlug = createAction<string>(
  "SET_CURRENT_SLUG"
);

export const geocodeCurrentAppearanceAddress = createAction.async<{}, ILatLng>(
  "GEOCODE_CURRENT_ADDRESS"
);
