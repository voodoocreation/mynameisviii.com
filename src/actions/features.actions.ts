import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("FEATURES");

export const addFeatures = actionCreator<PLAddFeatures>("ADD");

export const removeFeatures = actionCreator<PLRemoveFeatures>("REMOVE");
