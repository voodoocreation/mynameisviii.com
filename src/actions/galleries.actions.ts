import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("GALLERIES");

export const fetchGalleries = actionCreator.async<
  {},
  PLFetchGalleriesDone,
  PLFetchGalleriesFailed
>("FETCH");

export const fetchMoreGalleries = actionCreator.async<
  {},
  PLFetchGalleriesDone,
  PLFetchGalleriesFailed
>("FETCH_MORE");

export const fetchGalleryBySlug = actionCreator.async<
  PLFetchGalleryBySlugStarted,
  PLFetchGalleryBySlugDone,
  PLFetchGalleryBySlugFailed
>("FETCH_BY_SLUG");

export const setCurrentGallerySlug = actionCreator<PLSetCurrentGallerySlug>(
  "SET_CURRENT_SLUG"
);
