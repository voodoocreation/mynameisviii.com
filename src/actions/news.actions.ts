import actionCreatorFactory from "typescript-fsa";

const actionCreator = actionCreatorFactory("NEWS");

export const fetchLatestNews = actionCreator.async<
  {},
  PLFetchLatestNewsDone,
  PLFetchLatestNewsFailed
>("FETCH");

export const fetchMoreLatestNews = actionCreator.async<
  {},
  PLFetchLatestNewsDone,
  PLFetchLatestNewsFailed
>("FETCH_MORE");

export const fetchNewsArticleBySlug = actionCreator.async<
  PLFetchNewsArticleBySlugStarted,
  PLFetchNewsArticleBySlugDone,
  PLFetchNewsArticleBySlugFailed
>("FETCH_BY_SLUG");

export const setCurrentNewsArticleSlug = actionCreator<
  PLSetCurrentNewsArticleSlug
>("SET_CURRENT_SLUG");
