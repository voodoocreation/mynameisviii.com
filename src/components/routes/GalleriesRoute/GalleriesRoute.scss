@import "../../../scss/variables", "../../../scss/mixins";

.GalleriesRoute {
  padding-bottom: $spacing;

  &--listings {
    display: flex;
    flex-flow: row wrap;
    margin: (-$spacing-half);
  }
}
