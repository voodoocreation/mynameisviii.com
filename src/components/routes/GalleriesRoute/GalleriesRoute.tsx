import cn from "classnames";
import Head from "next/head";
import * as React from "react";
import { FormattedMessage, InjectedIntl } from "react-intl";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { ActionCreator } from "typescript-fsa";

import injectIntlIntoPage from "../../../helpers/injectIntlIntoPage";
import { absUrl } from "../../../transformers/transformData";
import OfflineNotice from "../../connected/OfflineNotice/OfflineNotice";
import ButtonBar from "../../presentation/ButtonBar/ButtonBar";
import GalleryListing from "../../presentation/GalleryListing/GalleryListing";
import LoadButton from "../../presentation/LoadButton/LoadButton";
import NoResults from "../../presentation/NoResults/NoResults";
import PageHeader from "../../presentation/PageHeader/PageHeader";

import * as actions from "../../../actions/root.actions";
import * as selectors from "../../../selectors/root.selectors";

interface IStoreProps {
  error?: IError;
  galleries: IGallery[];
  galleriesCount: number;
  hasAllGalleries: boolean;
  isLoading: boolean;
}

interface IDispatchProps {
  fetchGalleries: ActionCreator<{}>;
  fetchMoreGalleries: ActionCreator<{}>;
}

interface IProps extends IStoreProps, IDispatchProps {
  intl: InjectedIntl;
}

interface IState {
  loadedListings: {
    [index: string]: boolean;
  };
}

class GalleriesRoute extends React.Component<IProps, IState> {
  public static async getInitialProps(props: any) {
    const { isServer, store } = props.ctx;

    const state = store.getState();

    if (isServer || selectors.getGalleriesCount(state) === 0) {
      store.dispatch(actions.fetchGalleries.started({}));
    } else if (!selectors.getHasAllGalleries(state)) {
      store.dispatch(actions.fetchMoreGalleries.started({}));
    }
  }

  public readonly state = {
    loadedListings: {}
  };

  public render() {
    const { galleries, galleriesCount, error, hasAllGalleries } = this.props;
    const { formatMessage } = this.props.intl;

    const hasLoadedAllListings =
      galleriesCount === Object.keys(this.state.loadedListings).length;

    const isLoading = this.props.isLoading || !hasLoadedAllListings;

    const loadMoreButton = hasAllGalleries ? null : (
      <LoadButton
        isLoading={isLoading}
        isScrollLoadEnabled={!error}
        onLoad={this.onLoadMore}
      >
        {!error ? (
          <FormattedMessage id="LOAD_MORE" />
        ) : (
          <FormattedMessage id="TRY_AGAIN" />
        )}
      </LoadButton>
    );

    const pageTitle = formatMessage({ id: "GALLERIES_TITLE" });
    const pageDescription = formatMessage({ id: "GALLERIES_DESCRIPTION" });

    return (
      <article className={cn("GalleriesRoute", { hasLoadedAllListings })}>
        <Head>
          <title>
            {pageTitle}
            {" Â· "}
            {formatMessage({ id: "BRAND_NAME" })}
          </title>

          <meta content={pageDescription} name="description" />

          <meta property="og:title" content={pageTitle} />
          <meta property="og:description" content={pageDescription} />
          <meta property="og:type" content="website" />
          <meta
            property="og:image"
            content="https://s3.amazonaws.com/mynameisviii-static/galleries-og.jpg"
          />
          <meta property="og:url" content={absUrl("/galleries")} />
        </Head>

        <PageHeader>
          <FormattedMessage id="GALLERIES_TITLE" />
        </PageHeader>

        {error ? <OfflineNotice /> : null}

        {galleriesCount > 0 ? (
          <section className="GalleryListings">
            {galleries.map(gallery => (
              <GalleryListing
                {...gallery}
                key={gallery.slug}
                onLoad={this.onListingLoad}
              />
            ))}
          </section>
        ) : null}

        {hasAllGalleries && galleriesCount === 0 ? (
          <NoResults>
            <p>
              <FormattedMessage id="NO_GALLERIES" />
            </p>
          </NoResults>
        ) : null}

        <ButtonBar>{loadMoreButton}</ButtonBar>
      </article>
    );
  }

  private onListingLoad = (slug: string) => {
    this.setState({
      loadedListings: {
        ...this.state.loadedListings,
        [slug]: true
      }
    });
  };

  private onLoadMore = () => {
    this.props.fetchMoreGalleries({});
  };
}

const mapStateToProps = (state: any) => ({
  error: selectors.getGalleriesError(state),
  galleries: selectors.getGalleriesAsArray(state),
  galleriesCount: selectors.getGalleriesCount(state),
  hasAllGalleries: selectors.getHasAllGalleries(state),
  isLoading: selectors.getGalleriesIsLoading(state)
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      fetchGalleries: actions.fetchGalleries.started,
      fetchMoreGalleries: actions.fetchMoreGalleries.started
    },
    dispatch
  );

export default injectIntlIntoPage(
  connect<IStoreProps, IDispatchProps>(
    mapStateToProps,
    mapDispatchToProps
  )(GalleriesRoute)
);
