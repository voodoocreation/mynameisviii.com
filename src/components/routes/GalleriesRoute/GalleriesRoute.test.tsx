import * as actions from "../../../actions/root.actions";
import { gallery } from "../../../models/root.models";
import MockPageContext from "../../../utilities/MockPageContext";
import WrapperWithRedux from "../../../utilities/WrapperWithRedux";
import GalleriesRoute from "./GalleriesRoute";

const item1 = gallery({ slug: "test-1" });
const item2 = gallery({ slug: "test-2" });

const component = new WrapperWithRedux(GalleriesRoute);

describe("[routes] <GalleriesRoute />", () => {
  describe("getInitialProps", () => {
    const context = new MockPageContext();

    it("dispatches actions.fetchGalleries.started when rendering on the server", async () => {
      await GalleriesRoute.getInitialProps(context.toObject(true));

      expect(
        context.reduxHistory.filter(actions.fetchGalleries.started.match)
      ).toHaveLength(1);
    });

    it("dispatches actions.fetchGalleries.started when the store is empty", async () => {
      await GalleriesRoute.getInitialProps(context.toObject());

      expect(
        context.reduxHistory.filter(actions.fetchGalleries.started.match)
      ).toHaveLength(1);
    });

    it("dispatches actions.fetchMoreGalleries.started when there are more items to fetch", async () => {
      await GalleriesRoute.getInitialProps(
        context
          .withReduxState({
            galleries: {
              hasAllItems: false,
              items: {
                [item1.slug]: item1
              }
            }
          })
          .toObject()
      );

      expect(
        context.reduxHistory.filter(actions.fetchMoreGalleries.started.match)
      ).toHaveLength(1);
    });

    it("doesn't fetch anything when all items are in the store", async () => {
      await GalleriesRoute.getInitialProps(
        context
          .withReduxState({
            galleries: {
              hasAllItems: true,
              items: {
                [item1.slug]: item1
              }
            }
          })
          .toObject()
      );

      expect(
        context.reduxHistory.filter(actions.fetchGalleries.started.match)
      ).toHaveLength(0);
      expect(
        context.reduxHistory.filter(actions.fetchMoreGalleries.started.match)
      ).toHaveLength(0);
    });
  });

  describe("when all data is in the store", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          hasAllItems: true,
          items: {
            [item1.slug]: item1,
            [item2.slug]: item2
          }
        }
      })
      .mount();

    it("renders the listings section", () => {
      expect(wrapper.find(".GalleriesRoute--listings")).toHaveLength(1);
    });

    it("renders all listings", () => {
      expect(wrapper.find("GalleryListing")).toHaveLength(2);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe("when no data is in the store", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          hasAllItems: true,
          items: {}
        }
      })
      .mount();

    it("doesn't render the listings section", () => {
      expect(wrapper.find(".GalleriesRoute--listings")).toHaveLength(0);
    });

    it("renders the no results message", () => {
      expect(wrapper.find("NoResults")).toHaveLength(1);
    });
  });

  it("renders the offline notice when there's an error", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          hasError: true
        }
      })
      .mount();

    expect(wrapper.find("OfflineNotice")).toHaveLength(1);
  });

  it("renders with the 'load more' button when hasAllItems is false", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          hasAllItems: false,
          items: {
            [item1.slug]: item1
          }
        }
      })
      .mount();

    expect(wrapper.find("LoadButton")).toHaveLength(1);
  });

  it("renders without the 'load more' button when hasAllItems is true", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          hasAllItems: true
        }
      })
      .mount();

    expect(wrapper.find("LoadButton")).toHaveLength(0);
  });

  it("loads more listings when the 'load more' button is clicked", () => {
    const wrapper = component.mount();

    wrapper.find("LoadButton").simulate("click");

    expect(
      component.reduxHistory.filter(actions.fetchMoreGalleries.started.match)
    );
  });

  it("updates `loadedListings` state after a listing has loaded", () => {
    const wrapper = component
      .withReduxState({
        galleries: {
          items: {
            [item1.slug]: item1
          }
        }
      })
      .mount();

    // @ts-ignore-next-line
    wrapper.find("GalleryListing").prop("onLoad")();

    expect(wrapper.update().find(".hasLoadedAllListings")).toHaveLength(1);
  });
});
