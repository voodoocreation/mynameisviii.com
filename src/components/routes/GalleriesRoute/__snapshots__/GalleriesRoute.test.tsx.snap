// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[routes] <GalleriesRoute /> renders correctly 1`] = `
<article
  class="GalleriesRoute"
>
  <header
    class="PageHeader"
  >
    <h1>
      <span>
        Galleries
      </span>
    </h1>
  </header>
  <section
    class="GalleryListings"
  >
    <article
      class="GalleryListing"
    >
      <a
        href="/galleries/dark-spaces-i"
      >
        <header
          class="GalleryListing-header"
        >
          <h2>
            Dark Spaces I
          </h2>
        </header>
        <figure
          class="Image GalleryListing-image isLoading"
        >
          <div
            class="Image-wrapper"
          >
            <img
              alt="Dark Spaces I"
              src="https://s3.amazonaws.com/mynameisviii-static/galleries/dark-spaces-i/04.jpg"
            />
            <span
              class="Loader"
            >
              <div
                class="Loader-animation"
              >
                <span />
                <span />
                <span />
              </div>
            </span>
          </div>
        </figure>
      </a>
    </article>
    <article
      class="GalleryListing"
    >
      <a
        href="/galleries/test-gallery"
      >
        <header
          class="GalleryListing-header"
        >
          <h2 />
        </header>
      </a>
    </article>
  </section>
</article>
`;

exports[`[routes] <GalleriesRoute /> renders with an error correctly 1`] = `
<article
  class="GalleriesRoute hasLoadedAllListings"
>
  <header
    class="PageHeader"
  >
    <h1>
      <span>
        Galleries
      </span>
    </h1>
  </header>
  <div
    class="ButtonBar"
  >
    <button
      class="Button LoadButton"
      type="button"
    >
      <span
        class="Button-content"
      >
        <span>
          Try again
        </span>
      </span>
    </button>
  </div>
</article>
`;
